Topic/focus on tourer with zoom

danhenderson opened the pull request:
Added new props to Map component
- focusOnTourer - takes the id of the tourer you would like to centre in the map
- zoom - ability to set the initial zoom level when focusing on the tourer (default 10)

danhenderson committed:
1.0.1-15

danhenderson committed:
Add new props to allow to focus on a particular tourer at a specified zoom

danhenderson committed:
Merged in latest

bradparker commented on examples/source/waypoints/index.js:
```diff
@@ -98,6 +98,8 @@ const WaypointsExample = ({
       <TourTracker
         tourers={decoratedTourers}
         selected={selected}
+        focusOnTourer={decoratedTourers[0].id}
```
Why not use selected? Can you explain the behaviour this gives you?
Does it avoid unnecessary zooms on re-renders?
Do we think it'll get annoying?

Also could we follow the same sort of naming convention here: `focusedTourer`?

... in saying that it feels very much like `selected` fits that bill.

danhenderson replied:
Ok cool, will add tests.

As for selected, I hadn't really thought of using selected as I had sort of separated selected and focused in my head. But it probably makes sense to use selected.

The one "issue" is that I will still need to add some sort of flag to say focus on the selected tourer? In most cases, we still want to focus on the route regardless of who is selected, so I will still need another flag to say "I want to focus on the selected tourer"

bradparker reviewed (changes requested):
The approach is nice and neat... it'd be super nice if you could add a couple of specs to this as well?

Just assert that fitBounds is being called appropriately.

danhenderson committed:
Added distance to fetch route action payload

danhenderson committed:
Ability to connect routes

danhenderson committed:
Use selected to focus

danhenderson committed:
Added tests for fitBounds calls

danhenderson commented:
I have added a couple of extra requirements into this PR (probably should have kept them separate).

Updates to this morning's points:
- Now uses selected instead of the focusOnTourer prop
- Requires the focusMode prop to be set to 'selected' to focus on the selected tourer, otherwise it just focuses on the route as previous (not 100% sure about this bit)
- Added tests for the various fitBounds calls
- Also restructured the createSampleMap function slightly to take an object as it's only param, just so I could easily add and override certain props that are passed in

New points:
- connectRoutes prop which will render a line connecting cross continent routes
- the above has default styles (connectorDefaultStyles), but can be overriden using the connectorStyles prop
- the fetchRoute action now also includes the distance of each route in the payload if required (it is for S4S)

bradparker commented on source/actions/index.js:
```diff
@@ -12,9 +12,9 @@ const requestRoute = (index) => ({
 })
 
-const receiveRouteSuccess = (index, points) => ({
+const receiveRouteSuccess = (index, { points, distance }) => ({
```
Is this not the distance of the last point in the route?

danhenderson replied:
Correct. For some reason, I thought the distance in each of those points was the distance of that particular segment, rather than the distance from the start. I shall remove.

bradparker commented on source/actions/index.js:
```diff
@@ -25,8 +25,8 @@ const receiveRouteFailure = (index, error) => ({
 
-  return findRoute({ waypoints }).then((points) => (
-    dispatch(receiveRouteSuccess(index, points))
+  return findRoute({ waypoints }).then(route => (
```
`(parensPlz) => {}`

bradparker commented:
I'm not thrilled about focusMode. But I think I get the use case?
Without a selected rider (on load) – you want to zoom to the route... ?

bradparker commented on source/components/Map/index.js:
```diff
@@ -254,14 +270,21 @@ class Map extends React.Component {
     this._markers.addLayer(marker)
 
+    if (focusMode === 'selected' && tourer.id === selected) {
+      console.log('fitBounds on Selected')
```
whoops :P

danhenderson commented:
Basically, if focusMode is not set, then the map will focus on the route using the existing solution. But if it is set to 'selected', then it will focus on the selected rider. I chose a string in case someone might want to add 'selected-waypoint' or something similar in the future.

I too am not really a fan of it, but I am just not sure how else to handle the use case, where there is a selected rider, but you still want it to focus on the route as a whole as it does now. Or do you think it is acceptable that it always focuses on the selected rider by default?

bradparker commented:
Should be easy enough to remove the prop (make it more automatic) should we want to.
I guess it just pushes defining the behaviour out to the "host" application.

Little fixups and then I'm happy to merge.

danhenderson committed:
Added parens, remove debugging, and removed distance from action payload

danhenderson committed:
Formatting changes
